File: sdk/typescript/src/modules/treasury.ts
/**
 * Treasury Module
 *
 * Manage treasury spend proposals, budgets, and fund allocations.
 * All spend proposals are subject to governance timelock.
 */

import type { Coin } from '@cosmjs/stargate';
import type { StatesetClient } from '../client';
import type {
  SpendProposal,
  Budget,
  Allocation,
  TreasuryStats,
  TreasuryParams,
  ProposeSpendParams,
  SetBudgetParams,
  TxResult,
  PaginationParams,
} from '../types';
import { BUDGET_CATEGORIES } from '../constants';

export class TreasuryModule {
  constructor(private readonly client: StatesetClient) {}

  // ============================================================================
  // Queries
  // ============================================================================

  /**
   * Get treasury balance
   */
  async getBalance(): Promise<Coin[]> {
    const result = await this.client.queryModule('treasury', 'balance', {}) as {
      balance: Coin[];
    };
    return result.balance || [];
  }

  /**
   * Get treasury parameters
   */
  async getParams(): Promise<TreasuryParams> {
    const result = await this.client.queryModule('treasury', 'params', {}) as {
      params: TreasuryParams;
    };
    return result.params;
  }

  /**
   * Get treasury statistics
   */
  async getStats(): Promise<TreasuryStats> {
    const result = await this.client.queryModule('treasury', 'stats', {}) as {
      balance: Coin[];
      total_allocated: Coin[];
      total_disbursed: Coin[];
      pending_proposals: number;
      executed_proposals: number;
    };
    return {
      balance: result.balance,
      totalAllocated: result.total_allocated,
      totalDisbursed: result.total_disbursed,
      pendingProposals: result.pending_proposals,
      executedProposals: result.executed_proposals,
    };
  }

  /**
   * Get a spend proposal by ID
   */
  async getProposal(proposalId: string): Promise<SpendProposal | null> {
    try {
      const result = await this.client.queryModule('treasury', 'proposal', {
        proposal_id: proposalId,
      });
      return this.parseProposal(result);
    } catch {
      return null;
    }
  }

  /**
   * Get all spend proposals
   */
  async getProposals(
    status?: string,
    pagination?: PaginationParams,
  ): Promise<SpendProposal[]> {
    const result = await this.client.queryModule('treasury', 'proposals', {
      status,
      ...pagination,
    }) as { proposals: unknown[] };
    return (result.proposals || []).map((p) => this.parseProposal(p));
  }

  /**
   * Get pending proposals that can be executed
   */
  async getExecutableProposals(): Promise<SpendProposal[]> {
    const result = await this.client.queryModule(
      'treasury',
      'executable_proposals',
      {},
    ) as { proposals: unknown[] };
    return (result.proposals || []).map((p) => this.parseProposal(p));
  }

  /**
   * Get budget for a category
   */
  async getBudget(category: string): Promise<Budget | null> {
    try {
      const result = await this.client.queryModule('treasury', 'budget', {
        category,
      });
      return this.parseBudget(result);
    } catch {
      return null;
    }
  }

  /**
   * Get all budgets
   */
  async getAllBudgets(): Promise<Budget[]> {
    const result = await this.client.queryModule('treasury', 'budgets', {}) as {
      budgets: unknown[];
    };
    return (result.budgets || []).map((b) => this.parseBudget(b));
  }

  /**
   * Get allocation for a recipient
   */
  async getAllocation(recipient: string): Promise<Allocation | null> {
    try {
      const result = await this.client.queryModule('treasury', 'allocation', {
        recipient,
      });
      return this.parseAllocation(result);
    } catch {
      return null;
    }
  }

  /**
   * Check if a category is valid
   */
  isValidCategory(category: string): boolean {
    return BUDGET_CATEGORIES.includes(category as typeof BUDGET_CATEGORIES[number]);
  }

  // ============================================================================
  // Transactions
  // ============================================================================

  /**
   * Create a spend proposal (requires authority)
   */
  async proposeSpend(params: ProposeSpendParams): Promise<TxResult & { proposalId: string }> {
    if (!this.isValidCategory(params.category)) {
      throw new Error(`Invalid category: ${params.category}`);
    }

    const msg = {
      typeUrl: '/stateset.treasury.MsgProposeSpend',
      value: {
        authority: this.client.getAddress(),
        recipient: params.recipient,
        amount: params.amount,
        category: params.category,
        description: params.description,
        timelockSeconds: params.timelockSeconds.toString(),
      },
    };

    const result = await this.client.sendTx([msg], `Treasury spend: ${params.description}`);
    return {
      ...result,
      proposalId: result.transactionHash.slice(0, 16),
      events: [],
    };
  }

  /**
   * Execute a spend proposal after timelock expires
   */
  async executeSpend(proposalId: string): Promise<TxResult> {
    const msg = {
      typeUrl: '/stateset.treasury.MsgExecuteSpend',
      value: {
        authority: this.client.getAddress(),
        proposalId,
      },
    };

    const result = await this.client.sendTx([msg], 'Execute treasury spend');
    return { ...result, events: [] };
  }

  /**
   * Cancel a pending spend proposal
   */
  async cancelSpend(proposalId: string, reason: string): Promise<TxResult> {
    const msg = {
      typeUrl: '/stateset.treasury.MsgCancelSpend',
      value: {
        authority: this.client.getAddress(),
        proposalId,
        reason,
      },
    };

    const result = await this.client.sendTx([msg], 'Cancel treasury spend');
    return { ...result, events: [] };
  }

  /**
   * Set budget for a category (requires authority)
   */
  async setBudget(params: SetBudgetParams): Promise<TxResult> {
    if (!this.isValidCategory(params.category)) {
      throw new Error(`Invalid category: ${params.category}`);
    }

    const msg = {
      typeUrl: '/stateset.treasury.MsgSetBudget',
      value: {
        authority: this.client.getAddress(),
        category: params.category,
        totalLimit: params.totalLimit,
        periodLimit: params.periodLimit,
        periodDuration: (params.periodDurationDays * 24 * 60 * 60 * 1e9).toString(), // nanoseconds
        enabled: params.enabled,
      },
    };

    const result = await this.client.sendTx([msg], `Set budget: ${params.category}`);
    return { ...result, events: [] };
  }

  /**
   * Update treasury parameters (requires authority)
   */
  async updateParams(params: Partial<TreasuryParams>): Promise<TxResult> {
    const currentParams = await this.getParams();
    const newParams = { ...currentParams, ...params };

    const msg = {
      typeUrl: '/stateset.treasury.MsgUpdateParams',
      value: {
        authority: this.client.getAddress(),
        params: newParams,
      },
    };

    const result = await this.client.sendTx([msg], 'Update treasury params');
    return { ...result, events: [] };
  }

  // ============================================================================
  // Helpers
  // ============================================================================

  /**
   * Calculate time until a proposal can be executed
   */
  async getTimeUntilExecutable(proposalId: string): Promise<number> {
    const proposal = await this.getProposal(proposalId);
    if (!proposal) {
      throw new Error(`Proposal ${proposalId} not found`);
    }

    const now = Date.now();
    const executeAfter = proposal.executeAfter.getTime();
    return Math.max(0, executeAfter - now);
  }

  /**
   * Check if a proposal can be executed now
   */
  async canExecute(proposalId: string): Promise<boolean> {
    const proposal = await this.getProposal(proposalId);
    if (!proposal || proposal.status !== 'pending') {
      return false;
    }

    const now = Date.now();
    return (
      now >= proposal.executeAfter.getTime() &&
      now < proposal.expiresAt.getTime()
    );
  }

  private parseProposal(data: unknown): SpendProposal {
    const p = data as {
      id: string;
      proposer: string;
      recipient: string;
      amount: Coin[];
      category: string;
      description: string;
      status: string;
      created_at: string;
      execute_after: string;
      expires_at: string;
      executed_at?: string;
    };
    return {
      id: p.id,
      proposer: p.proposer,
      recipient: p.recipient,
      amount: p.amount,
      category: p.category,
      description: p.description,
      status: p.status as SpendProposal['status'],
      createdAt: new Date(p.created_at),
      executeAfter: new Date(p.execute_after),
      expiresAt: new Date(p.expires_at),
      executedAt: p.executed_at ? new Date(p.executed_at) : undefined,
    };
  }

  private parseBudget(data: unknown): Budget {
    const b = data as {
      category: string;
      total_limit: Coin[];
      period_limit: Coin[];
      period_spent: Coin[];
      total_spent: Coin[];
      period_start: string;
      period_duration: string;
      enabled: boolean;
    };
    return {
      category: b.category,
      totalLimit: b.total_limit,
      periodLimit: b.period_limit,
      periodSpent: b.period_spent,
      totalSpent: b.total_spent,
      periodStart: new Date(b.period_start),
      periodDuration: parseInt(b.period_duration, 10),
      enabled: b.enabled,
    };
  }

  private parseAllocation(data: unknown): Allocation {
    const a = data as {
      recipient: string;
      total_allocated: Coin[];
      total_disbursed: Coin[];
      pending: Coin[];
      last_updated: string;
    };
    return {
      recipient: a.recipient,
      totalAllocated: a.total_allocated,
      totalDisbursed: a.total_disbursed,
      pending: a.pending,
      lastUpdated: new Date(a.last_updated),
    };
  }
}

File: x/treasury/types/treasury.go
package types

import (
	"time"

	errorsmod "cosmossdk.io/errors"
	sdkmath "cosmossdk.io/math"
	sdk "github.com/cosmos/cosmos-sdk/types"
)

// TreasuryParams defines the parameters for the treasury module
type TreasuryParams struct {
	// MinTimelockDuration is the minimum timelock for spend proposals (e.g., 24 hours)
	MinTimelockDuration time.Duration `json:"min_timelock_duration" yaml:"min_timelock_duration"`
	// MaxTimelockDuration is the maximum timelock for spend proposals (e.g., 30 days)
	MaxTimelockDuration time.Duration `json:"max_timelock_duration" yaml:"max_timelock_duration"`
	// ProposalExpiryDuration is how long a proposal remains valid after timelock expires
	ProposalExpiryDuration time.Duration `json:"proposal_expiry_duration" yaml:"proposal_expiry_duration"`
	// MaxPendingProposals is the maximum number of pending proposals allowed
	MaxPendingProposals uint32 `json:"max_pending_proposals" yaml:"max_pending_proposals"`
	// EmergencyMultisigThreshold for bypassing timelock (requires N of M signers)
	EmergencyMultisigThreshold uint32 `json:"emergency_multisig_threshold" yaml:"emergency_multisig_threshold"`
	// BaseBurnRate percentage of incoming fees to burn (basis points, e.g., 2500 = 25%)
	BaseBurnRate uint32 `json:"base_burn_rate" yaml:"base_burn_rate"`
	// ValidatorRewardRate percentage to validators (basis points)
	ValidatorRewardRate uint32 `json:"validator_reward_rate" yaml:"validator_reward_rate"`
	// CommunityPoolRate percentage to community pool (basis points)
	CommunityPoolRate uint32 `json:"community_pool_rate" yaml:"community_pool_rate"`
}

// DefaultTreasuryParams returns default treasury parameters
func DefaultTreasuryParams() TreasuryParams {
	return TreasuryParams{
		MinTimelockDuration:        24 * time.Hour,        // 24 hours minimum
		MaxTimelockDuration:        30 * 24 * time.Hour,   // 30 days maximum
		ProposalExpiryDuration:     7 * 24 * time.Hour,    // 7 days to execute after timelock
		MaxPendingProposals:        100,
		EmergencyMultisigThreshold: 3,                     // 3 of 5 for emergency
		BaseBurnRate:               2500,                  // 25% burned
		ValidatorRewardRate:        5000,                  // 50% to validators
		CommunityPoolRate:          2500,                  // 25% to community pool
	}
}

// Validate validates treasury params
func (p TreasuryParams) Validate() error {
	if p.MinTimelockDuration < time.Hour {
		return errorsmod.Wrap(ErrTimelockTooShort, "minimum timelock must be at least 1 hour")
	}
	if p.MaxTimelockDuration < p.MinTimelockDuration {
		return errorsmod.Wrap(ErrTimelockTooShort, "max timelock must be >= min timelock")
	}
	totalRate := p.BaseBurnRate + p.ValidatorRewardRate + p.CommunityPoolRate
	if totalRate != 10000 {
		return errorsmod.Wrapf(ErrInvalidAmount, "fee distribution rates must sum to 10000 (100%%), got %d", totalRate)
	}
	return nil
}

// SpendProposal represents a time-locked treasury spend proposal
type SpendProposal struct {
	Id           uint64    `json:"id" yaml:"id"`
	Proposer     string    `json:"proposer" yaml:"proposer"`
	Recipient    string    `json:"recipient" yaml:"recipient"`
	Amount       sdk.Coins `json:"amount" yaml:"amount"`
	Category     string    `json:"category" yaml:"category"`
	Description  string    `json:"description" yaml:"description"`
	Status       string    `json:"status" yaml:"status"`
	CreatedAt    time.Time `json:"created_at" yaml:"created_at"`
	ExecuteAfter time.Time `json:"execute_after" yaml:"execute_after"`
	ExpiresAt    time.Time `json:"expires_at" yaml:"expires_at"`
	ExecutedAt   time.Time `json:"executed_at,omitempty" yaml:"executed_at,omitempty"`
	TxHash       string    `json:"tx_hash,omitempty" yaml:"tx_hash,omitempty"`
}

// ValidateBasic performs basic validation
func (p SpendProposal) ValidateBasic() error {
	if _, err := sdk.AccAddressFromBech32(p.Proposer); err != nil {
		return errorsmod.Wrap(ErrUnauthorized, "invalid proposer address")
	}
	if _, err := sdk.AccAddressFromBech32(p.Recipient); err != nil {
		return errorsmod.Wrap(ErrInvalidRecipient, "invalid recipient address")
	}
	if !p.Amount.IsValid() || p.Amount.IsZero() {
		return errorsmod.Wrap(ErrInvalidAmount, "amount must be valid and non-zero")
	}
	if !IsValidCategory(p.Category) {
		return errorsmod.Wrapf(ErrInvalidCategory, "unknown category: %s", p.Category)
	}
	if len(p.Description) == 0 {
		return errorsmod.Wrap(ErrInvalidAmount, "description cannot be empty")
	}
	if len(p.Description) > 1000 {
		return errorsmod.Wrap(ErrInvalidAmount, "description too long (max 1000 chars)")
	}
	return nil
}

// CanExecute checks if the proposal can be executed now
func (p SpendProposal) CanExecute(now time.Time) error {
	if p.Status != SpendStatusPending {
		return ErrProposalNotPending
	}
	if now.Before(p.ExecuteAfter) {
		return ErrTimelockNotExpired
	}
	if now.After(p.ExpiresAt) {
		return ErrProposalExpired
	}
	return nil
}

// Budget represents a category budget with limits
type Budget struct {
	Category       string         `json:"category" yaml:"category"`
	TotalLimit     sdk.Coins      `json:"total_limit" yaml:"total_limit"`
	PeriodLimit    sdk.Coins      `json:"period_limit" yaml:"period_limit"`
	PeriodSpent    sdk.Coins      `json:"period_spent" yaml:"period_spent"`
	TotalSpent     sdk.Coins      `json:"total_spent" yaml:"total_spent"`
	PeriodStart    time.Time      `json:"period_start" yaml:"period_start"`
	PeriodDuration time.Duration  `json:"period_duration" yaml:"period_duration"`
	Enabled        bool           `json:"enabled" yaml:"enabled"`
}

// ValidateBasic performs basic validation
func (b Budget) ValidateBasic() error {
	if !IsValidCategory(b.Category) {
		return errorsmod.Wrapf(ErrInvalidCategory, "unknown category: %s", b.Category)
	}
	if !b.TotalLimit.IsValid() {
		return errorsmod.Wrap(ErrInvalidAmount, "invalid total limit")
	}
	if !b.PeriodLimit.IsValid() {
		return errorsmod.Wrap(ErrInvalidAmount, "invalid period limit")
	}
	return nil
}

// CanSpend checks if the budget allows spending the amount
func (b Budget) CanSpend(amount sdk.Coins, now time.Time) error {
	if !b.Enabled {
		return errorsmod.Wrapf(ErrBudgetExceeded, "budget for %s is disabled", b.Category)
	}

	// Check period limit
	periodSpent := b.PeriodSpent
	if now.After(b.PeriodStart.Add(b.PeriodDuration)) {
		// New period, reset spent
		periodSpent = sdk.NewCoins()
	}

	newPeriodSpent := periodSpent.Add(amount...)
	for _, coin := range newPeriodSpent {
		limit := b.PeriodLimit.AmountOf(coin.Denom)
		if coin.Amount.GT(limit) {
			return errorsmod.Wrapf(ErrBudgetExceeded, "period limit exceeded for %s in category %s", coin.Denom, b.Category)
		}
	}

	// Check total limit
	newTotalSpent := b.TotalSpent.Add(amount...)
	for _, coin := range newTotalSpent {
		limit := b.TotalLimit.AmountOf(coin.Denom)
		if !limit.IsZero() && coin.Amount.GT(limit) {
			return errorsmod.Wrapf(ErrBudgetExceeded, "total limit exceeded for %s in category %s", coin.Denom, b.Category)
		}
	}

	return nil
}

// Allocation tracks funds allocated to a specific recipient
type Allocation struct {
	Recipient      string    `json:"recipient" yaml:"recipient"`
	TotalAllocated sdk.Coins `json:"total_allocated" yaml:"total_allocated"`
	TotalDisbursed sdk.Coins `json:"total_disbursed" yaml:"total_disbursed"`
	Pending        sdk.Coins `json:"pending" yaml:"pending"`
	LastUpdated    time.Time `json:"last_updated" yaml:"last_updated"`
}

// RevenueRecord tracks incoming revenue to the treasury
type RevenueRecord struct {
	Id        uint64    `json:"id" yaml:"id"`
	Source    string    `json:"source" yaml:"source"`
	Amount    sdk.Coins `json:"amount" yaml:"amount"`
	Timestamp time.Time `json:"timestamp" yaml:"timestamp"`
	TxHash    string    `json:"tx_hash" yaml:"tx_hash"`
	Metadata  string    `json:"metadata" yaml:"metadata"`
}

// TreasuryStats provides aggregate statistics
type TreasuryStats struct {
	TotalBalance         sdk.Coins              `json:"total_balance" yaml:"total_balance"`
	TotalAllocated       sdk.Coins              `json:"total_allocated" yaml:"total_allocated"`
	TotalDisbursed       sdk.Coins              `json:"total_disbursed" yaml:"total_disbursed"`
	TotalBurned          sdk.Coins              `json:"total_burned" yaml:"total_burned"`
	TotalRevenue         sdk.Coins              `json:"total_revenue" yaml:"total_revenue"`
	PendingProposals     uint32                 `json:"pending_proposals" yaml:"pending_proposals"`
	ExecutedProposals    uint32                 `json:"executed_proposals" yaml:"executed_proposals"`
	CategorySpending     map[string]sdk.Coins   `json:"category_spending" yaml:"category_spending"`
	LastRevenueTimestamp time.Time              `json:"last_revenue_timestamp" yaml:"last_revenue_timestamp"`
}

// IsValidCategory checks if the category is valid
func IsValidCategory(category string) bool {
	switch category {
	case CategoryDevelopment, CategoryMarketing, CategoryOperations,
		CategoryGrants, CategorySecurity, CategoryInfrastructure, CategoryReserve:
		return true
	}
	return false
}

// ValidCategories returns all valid budget categories
func ValidCategories() []string {
	return []string{
		CategoryDevelopment,
		CategoryMarketing,
		CategoryOperations,
		CategoryGrants,
		CategorySecurity,
		CategoryInfrastructure,
		CategoryReserve,
	}
}

// FeeDistribution represents how fees are distributed
type FeeDistribution struct {
	BurnAmount          sdk.Coins `json:"burn_amount" yaml:"burn_amount"`
	ValidatorAmount     sdk.Coins `json:"validator_amount" yaml:"validator_amount"`
	CommunityPoolAmount sdk.Coins `json:"community_pool_amount" yaml:"community_pool_amount"`
	TreasuryAmount      sdk.Coins `json:"treasury_amount" yaml:"treasury_amount"`
}

// CalculateFeeDistribution calculates how to distribute fees
func CalculateFeeDistribution(fees sdk.Coins, params TreasuryParams) FeeDistribution {
	dist := FeeDistribution{
		BurnAmount:          sdk.NewCoins(),
		ValidatorAmount:     sdk.NewCoins(),
		CommunityPoolAmount: sdk.NewCoins(),
		TreasuryAmount:      sdk.NewCoins(),
	}

	for _, coin := range fees {
		burnAmt := coin.Amount.Mul(sdkmath.NewInt(int64(params.BaseBurnRate))).Quo(sdkmath.NewInt(10000))
		validatorAmt := coin.Amount.Mul(sdkmath.NewInt(int64(params.ValidatorRewardRate))).Quo(sdkmath.NewInt(10000))
		communityAmt := coin.Amount.Mul(sdkmath.NewInt(int64(params.CommunityPoolRate))).Quo(sdkmath.NewInt(10000))

		if burnAmt.IsPositive() {
			dist.BurnAmount = dist.BurnAmount.Add(sdk.NewCoin(coin.Denom, burnAmt))
		}
		if validatorAmt.IsPositive() {
			dist.ValidatorAmount = dist.ValidatorAmount.Add(sdk.NewCoin(coin.Denom, validatorAmt))
		}
		if communityAmt.IsPositive() {
			dist.CommunityPoolAmount = dist.CommunityPoolAmount.Add(sdk.NewCoin(coin.Denom, communityAmt))
		}
	}

	return dist
}
